### PR TITLE
Keyboard mapping: add mapped key string to raw key event

### PR DESCRIPTION
**Pull request by yatli**

### CODE REVIEW

**Comment by yatli** [comment]
Getting the error: 

>  error : MembersMustExist : Member 'public void Avalonia.Input.Raw.RawKeyEventArgs..ctor(Avalonia.Input.IKeyboardDevice, System.UInt64, Avalonia.Input.IInputRoot, Avalonia.Input.Raw.RawKeyEventType, Avalonia.Input.Key, Avalonia.Input.RawInputModifiers)

**Comment by yatli** [comment]
OSX side not implemented. Don't really know how to bite the objc+com(?)+sharpgen thing #_# (sorry

**Comment by kekekeks** [comment]
How would that work with various IME systems once they are implemented?

**Comment by kekekeks** [comment]
Also, I don't see any the user API changes, only internal raw event args.

**Comment by yatli** [comment]
> Also, I don't see any the user API changes, only internal raw event args.

Yeah just realized that. Adding it to `KeyEventArgs` now.

**Comment by yatli** [comment]
> How would that work with various IME systems once they are implemented?

I think IME input events should be routed as text input messages plus `ImeProcessedKey` just like the Windows side.
For KeyEventArg they can be reported as raw keys, because the committed character isn't confirmed until the user commits the sequence (usually with space/enter etc.) anyway

**Comment by kekekeks** [comment]
I don't see such property in neither [WPF](https://docs.microsoft.com/en-us/dotnet/api/system.windows.input.keyeventargs?view=netcore-3.1) or [UWP](https://docs.microsoft.com/en-us/uwp/api/windows.ui.core.keyeventargs?view=winrt-19041), why do we need it?

**Comment by yatli** [comment]
see: https://github.com/yatli/fvim/issues/36

**Comment by maxkatz6** [comment]
@kekekeks uwp has Key and OriginalKey (unmapped) props https://docs.microsoft.com/en-us/uwp/api/windows.ui.xaml.input.keyroutedeventargs?view=winrt-19041

**Comment by maxkatz6** [comment]
And I have looked detailedly on original issue, but probably Character prop is needed as well https://docs.microsoft.com/en-us/uwp/api/windows.ui.xaml.input.characterreceivedroutedeventargs?view=winrt-19041 (note, it is from another event in uwp). Not sure, does Avalonia have it, can't check now.

File: src/Windows/Avalonia.Win32/WindowImpl.AppWndProc.cs [hunk 1]
```diff
@@ -103,13 +103,14 @@ protected virtual unsafe IntPtr AppWndProc(IntPtr hWnd, uint msg, IntPtr wParam,
...
-                            RawKeyEventType.KeyDown,
-                            KeyInterop.KeyFromVirtualKey(ToInt32(wParam), ToInt32(lParam)),
-                            WindowsKeyboardDevice.Instance.Modifiers);
+                    e = new RawKeyEventArgs(
+                        WindowsKeyboardDevice.Instance,
+                        timestamp,
+                        _owner,
+                        RawKeyEventType.KeyDown,
+                        KeyInterop.KeyFromVirtualKey(ToInt32(wParam), ToInt32(lParam)),
+                        WindowsKeyboardDevice.Instance.StringFromVirtualKey((uint)ToInt32(wParam)),
```
Comment (yatli) [inline src/Windows/Avalonia.Win32/WindowImpl.AppWndProc.cs:112]: This needs to be corrected. `StringFromVirtualKey` takes `ctrl` and `alt` into consideration and will translate a key into a control sequence.

**Comment by kekekeks** [comment]
@maxkatz6 UWP seems to provide virtual key mapping, not characters.

**Comment by yatli** [comment]
@kekekeks I think I can give a single requirement that cannot be satisfied with the current framework: to detect `Control-#` on all keyboard mappings.

`TextInput` doesn't give key modifiers. You get "#"
`KeyEventArgs` doesn't give key mappings. You get "en-US key at # position"

**Comment by kekekeks** [comment]
Before introducing such an API that doesn't exist in UWP/WPF, we need to
1) check how UWP and WPF achieve the same result
2) check how it's done in Qt and GTK.

Otherwise we are bound to miss something important.

**Comment by yatli** [comment]
For Qt, see `neovim-qt`: https://github.com/equalsraf/neovim-qt/blob/4352846be7f7f42738c568eafaba965a20bb7a45/src/gui/input.cpp#L132

**Comment by grokys** [comment]
What's the status on this PR?

**Comment by workgroupengineering** [comment]
instead of modifying RawKeyEventArgs, isn't it better to create an extesion method like this?

```csharp
public interface IKeyMapper
{
   string ToKeyString(RawKeyEventArgs arg, CultureInfo culture)
   ...
}


public static string? ToKeyString(this RawKeyEventArgs args, CultureInfo culture = null, )
{
     var mapper = AvaloniaLocator.GetService<IKeyMapper>();
     return mapper?.ToKeyString(arg, culture);
}
```

**Comment by kekekeks** [comment]
1) Using locator in new code is heavily discouraged unless there is a **very** good reason why it can't be done in a different way
2) The information about key mappings is heavily platform-dependent and there is no guarantee that information from RawKeyEventEventArgs would be enough to get the mapped string.

**Comment by workgroupengineering** [comment]
> 
>     1. Using locator in new code is heavily discouraged unless there is a **very** good reason why it can't be done in a different way
Locator in this case is useful for injecting the specific keymapper by platform, we could implement macOsKeyMapper, iOSKeyMapper, etc ...

> 
>     2. The information about key mappings is heavily platform-dependent and there is no guarantee that information from RawKeyEventEventArgs would be enough to get the mapped string.

that's true, but any platform could subclass RawKeyEventArgs and add platform-specific information. Only the keymapper of the specific platform would see the difference.

**Comment by kekekeks** [comment]
>Locator in this case is useful for injecting the specific keymapper by platform, we could implement macOsKeyMapper, iOSKeyMapper, etc ...

Which is *not* a valid reason to add something to service locator.  Stuff like that should be resolved via the platform layer or embedded into the event args via the same platform layer. Service locator is pure evil that's here because it's been there in the early days (and it's also kinda troublesome to manage rendering-related stuff without a locator).

**Comment by workgroupengineering** [comment]
I think that there is no absolute evil or absolute good, it all depends on the use made of it.
In this case I don't think it is.
By linking the key string to the event, it would be much more complicated to retrieve key strings in a specific culture.

**Comment by maxkatz6** [comment]
> @maxkatz6 UWP seems to provide virtual key mapping, not characters.

Yep, you are right. I misleaded point of PR back then.

I just found that in UWP there is relatively new event for controls - [UIElement.CharacterReceived](https://docs.microsoft.com/en-us/uwp/api/windows.ui.xaml.uielement.characterreceived?view=winrt-19041)
Which has ```char Character { get; }``` in the args.

There is also old global event with another arguments [CoreWindow.CharacterReceived](https://docs.microsoft.com/en-us/uwp/api/windows.ui.core.corewindow.characterreceived?view=winrt-19041)

**Comment by yatli** [comment]
Another approach is to provide API to query "what text this keystroke will produce".
We just need to keep the relevant platform-specific keystroke data.
This could also play nicer with IME because we can get the info from the query that IME keystrokes are "queued", and will only produce text when it's committed.

**Comment by mcmikecreations** [comment]
Has any progress been done since the last message?

**Comment by yatli** [comment]
The core team has designed and implemented the IME model, I have to read it and adjust my proposal so not to break it :)

**Comment by mcmikecreations** [comment]
> 
> 
> The core team has designed and implemented the IME model, I have to read it and adjust my proposal so not to break it :)

I have the same problem as [this user of fvim](https://github.com/yatli/fvim/issues/36) and I'm looking forward to a fix that I can use in my project. Adding keyboard mapping would benefit a lot of projects!

**Comment by jmacato** [comment]
Closing this PR for inactivity. Ping me if this is updated accordingly.

**Pull request closed without merging**